linera-execution/src/execution_state_actor.rs

```diff
@@ -402,12 +403,25 @@ where
 
             ReadBlobContent { blob_id, callback } => {
                 let blob = self.system.read_blob_content(blob_id).await?;
+                if blob_id.blob_type == BlobType::Data {
+                    resource_controller
+                        .with_state(&mut self.system)
+                        .await?
+                        .track_blob_read(blob.bytes().len() as u64)?;
+                }
                 let is_new = self.system.blob_used(None, blob_id).await?;
                 callback.respond((blob, is_new))
             }
 
             AssertBlobExists { blob_id, callback } => {
                 self.system.assert_blob_exists(blob_id).await?;
+                // Treating this as reading a size-0 blob for fee purposes.
+                if blob_id.blob_type == BlobType::Data {
+                    resource_controller
+                        .with_state(&mut self.system)
+                        .await?
+                        .track_blob_read(0)?;
+                }
```

Comment:
This won't get recorded in the case when blob does not exist.

Reply:
Yes, then the whole block is invalid.